[RISCV] Can't build Google highway RVV code using builtin's on OpenBSD

Trying to build Google highway on RISCV using RVV it fails to build with errors regarding the use of builtin's.

Downstream seems to think this is an issue with Clang. I was able to build Ok using GCC.

Downstream also pointed out that OpenBSD using uint64_t as ```unsigned long long``` instead of ```unsigned long``` was the cause of troubles for Clang with the builtins.

```
openbsd-riscv64$ cmake --build .
[  1%] Building CXX object CMakeFiles/hwy.dir/hwy/per_target.cc.o
In file included from /home/brad/tmp/highway/hwy/per_target.cc:28:
In file included from /home/brad/tmp/highway/hwy/highway.h:714:
/home/brad/tmp/highway/hwy/ops/rvv-inl.h:2466:17: error: no matching function for call to '__riscv_vse64_v_u64m1'
 2466 | HWY_RVV_FOREACH(HWY_RVV_STOREN, StoreN, se, _ALL_VIRT)
      | ~~~~~~~~~~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/home/brad/tmp/highway/hwy/ops/rvv-inl.h:477:22: note: expanded from macro 'HWY_RVV_FOREACH'
  477 |   HWY_RVV_FOREACH_UI(X_MACRO, NAME, OP, LMULS)    \
      |   ~~~~~~~~~~~~~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~
/home/brad/tmp/highway/hwy/ops/rvv-inl.h:473:21: note: expanded from macro 'HWY_RVV_FOREACH_UI'
  473 |   HWY_RVV_FOREACH_U(X_MACRO, NAME, OP, LMULS)        \
      |   ~~~~~~~~~~~~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~
/home/brad/tmp/highway/hwy/ops/rvv-inl.h:461:27: note: expanded from macro 'HWY_RVV_FOREACH_U'
  461 |   HWY_RVV_FOREACH_U163264(X_MACRO, NAME, OP, LMULS)
      |   ~~~~~~~~~~~~~~~~~~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~
note: (skipping 5 expansions in backtrace; use -fmacro-backtrace-limit=0 to see all)
/home/brad/tmp/highway/hwy/ops/rvv-inl.h:213:3: note: expanded from macro 'HWY_RVV_FOREACH_64_LE2'
  213 |   X_MACRO(BASE, CHAR, 64, __, 32, m1, m2, mf2, 0, /*MLEN=*/64, NAME, OP) \
      |   ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/home/brad/tmp/highway/hwy/ops/rvv-inl.h:2463:12: note: expanded from macro 'HWY_RVV_STOREN'
 2463 |     return __riscv_v##OP##SEW##_v_##CHAR##SEW##LMUL(                           \
      |            ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
<scratch space>:98:1: note: expanded from here
   98 | __riscv_vse64_v_u64m1
      | ^~~~~~~~~~~~~~~~~~~~~
/home/brad/tmp/highway/hwy/ops/rvv-inl.h:2466:17: note: candidate function not viable: no known conversion from 'unsigned long long *' to 'unsigned long *' for 1st argument
 2466 | HWY_RVV_FOREACH(HWY_RVV_STOREN, StoreN, se, _ALL_VIRT)
      | ~~~~~~~~~~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/home/brad/tmp/highway/hwy/ops/rvv-inl.h:477:22: note: expanded from macro 'HWY_RVV_FOREACH'
  477 |   HWY_RVV_FOREACH_UI(X_MACRO, NAME, OP, LMULS)    \
      |   ~~~~~~~~~~~~~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~
/home/brad/tmp/highway/hwy/ops/rvv-inl.h:473:21: note: expanded from macro 'HWY_RVV_FOREACH_UI'
  473 |   HWY_RVV_FOREACH_U(X_MACRO, NAME, OP, LMULS)        \
      |   ~~~~~~~~~~~~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~
/home/brad/tmp/highway/hwy/ops/rvv-inl.h:461:27: note: expanded from macro 'HWY_RVV_FOREACH_U'
  461 |   HWY_RVV_FOREACH_U163264(X_MACRO, NAME, OP, LMULS)
      |   ~~~~~~~~~~~~~~~~~~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~
note: (skipping 5 expansions in backtrace; use -fmacro-backtrace-limit=0 to see all)
/home/brad/tmp/highway/hwy/ops/rvv-inl.h:213:3: note: expanded from macro 'HWY_RVV_FOREACH_64_LE2'
  213 |   X_MACRO(BASE, CHAR, 64, __, 32, m1, m2, mf2, 0, /*MLEN=*/64, NAME, OP) \
      |   ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/home/brad/tmp/highway/hwy/ops/rvv-inl.h:2463:12: note: expanded from macro 'HWY_RVV_STOREN'
 2463 |     return __riscv_v##OP##SEW##_v_##CHAR##SEW##LMUL(                           \
      |            ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
<scratch space>:98:1: note: expanded from here
   98 | __riscv_vse64_v_u64m1
      | ^~~~~~~~~~~~~~~~~~~~~
In file included from /home/brad/tmp/highway/hwy/per_target.cc:28:
In file included from /home/brad/tmp/highway/hwy/highway.h:714:
/home/brad/tmp/highway/hwy/ops/rvv-inl.h:2466:17: error: no matching function for call to '__riscv_vse64_v_u64m2'
 2466 | HWY_RVV_FOREACH(HWY_RVV_STOREN, StoreN, se, _ALL_VIRT)
      | ~~~~~~~~~~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/home/brad/tmp/highway/hwy/ops/rvv-inl.h:477:22: note: expanded from macro 'HWY_RVV_FOREACH'
  477 |   HWY_RVV_FOREACH_UI(X_MACRO, NAME, OP, LMULS)    \
      |   ~~~~~~~~~~~~~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~
/home/brad/tmp/highway/hwy/ops/rvv-inl.h:473:21: note: expanded from macro 'HWY_RVV_FOREACH_UI'
  473 |   HWY_RVV_FOREACH_U(X_MACRO, NAME, OP, LMULS)        \
      |   ~~~~~~~~~~~~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~
/home/brad/tmp/highway/hwy/ops/rvv-inl.h:461:27: note: expanded from macro 'HWY_RVV_FOREACH_U'
  461 |   HWY_RVV_FOREACH_U163264(X_MACRO, NAME, OP, LMULS)
      |   ~~~~~~~~~~~~~~~~~~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~
note: (skipping 5 expansions in backtrace; use -fmacro-backtrace-limit=0 to see all)
/home/brad/tmp/highway/hwy/ops/rvv-inl.h:214:3: note: expanded from macro 'HWY_RVV_FOREACH_64_LE2'
  214 |   X_MACRO(BASE, CHAR, 64, __, 32, m2, m4, m1, 1, /*MLEN=*/32, NAME, OP)
      |   ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/home/brad/tmp/highway/hwy/ops/rvv-inl.h:2463:12: note: expanded from macro 'HWY_RVV_STOREN'
 2463 |     return __riscv_v##OP##SEW##_v_##CHAR##SEW##LMUL(                           \
      |            ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
<scratch space>:112:1: note: expanded from here
  112 | __riscv_vse64_v_u64m2
      | ^~~~~~~~~~~~~~~~~~~~~
/home/brad/tmp/highway/hwy/ops/rvv-inl.h:2466:17: note: candidate function not viable: no known conversion from 'unsigned long long *' to 'unsigned long *' for 1st argument
 2466 | HWY_RVV_FOREACH(HWY_RVV_STOREN, StoreN, se, _ALL_VIRT)
      | ~~~~~~~~~~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/home/brad/tmp/highway/hwy/ops/rvv-inl.h:477:22: note: expanded from macro 'HWY_RVV_FOREACH'
  477 |   HWY_RVV_FOREACH_UI(X_MACRO, NAME, OP, LMULS)    \
      |   ~~~~~~~~~~~~~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~
/home/brad/tmp/highway/hwy/ops/rvv-inl.h:473:21: note: expanded from macro 'HWY_RVV_FOREACH_UI'
  473 |   HWY_RVV_FOREACH_U(X_MACRO, NAME, OP, LMULS)        \
      |   ~~~~~~~~~~~~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~
/home/brad/tmp/highway/hwy/ops/rvv-inl.h:461:27: note: expanded from macro 'HWY_RVV_FOREACH_U'
  461 |   HWY_RVV_FOREACH_U163264(X_MACRO, NAME, OP, LMULS)
      |   ~~~~~~~~~~~~~~~~~~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~
note: (skipping 5 expansions in backtrace; use -fmacro-backtrace-limit=0 to see all)
/home/brad/tmp/highway/hwy/ops/rvv-inl.h:214:3: note: expanded from macro 'HWY_RVV_FOREACH_64_LE2'
  214 |   X_MACRO(BASE, CHAR, 64, __, 32, m2, m4, m1, 1, /*MLEN=*/32, NAME, OP)
      |   ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/home/brad/tmp/highway/hwy/ops/rvv-inl.h:2463:12: note: expanded from macro 'HWY_RVV_STOREN'
 2463 |     return __riscv_v##OP##SEW##_v_##CHAR##SEW##LMUL(                           \
      |            ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
<scratch space>:112:1: note: expanded from here
  112 | __riscv_vse64_v_u64m2
      | ^~~~~~~~~~~~~~~~~~~~~
In file included from /home/brad/tmp/highway/hwy/per_target.cc:28:
In file included from /home/brad/tmp/highway/hwy/highway.h:714:
/home/brad/tmp/highway/hwy/ops/rvv-inl.h:2466:17: error: no matching function for call to '__riscv_vse64_v_u64m4'
 2466 | HWY_RVV_FOREACH(HWY_RVV_STOREN, StoreN, se, _ALL_VIRT)
      | ~~~~~~~~~~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/home/brad/tmp/highway/hwy/ops/rvv-inl.h:477:22: note: expanded from macro 'HWY_RVV_FOREACH'
  477 |   HWY_RVV_FOREACH_UI(X_MACRO, NAME, OP, LMULS)    \
      |   ~~~~~~~~~~~~~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~
/home/brad/tmp/highway/hwy/ops/rvv-inl.h:473:21: note: expanded from macro 'HWY_RVV_FOREACH_UI'
  473 |   HWY_RVV_FOREACH_U(X_MACRO, NAME, OP, LMULS)        \
      |   ~~~~~~~~~~~~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~
/home/brad/tmp/highway/hwy/ops/rvv-inl.h:461:27: note: expanded from macro 'HWY_RVV_FOREACH_U'
  461 |   HWY_RVV_FOREACH_U163264(X_MACRO, NAME, OP, LMULS)
      |   ~~~~~~~~~~~~~~~~~~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~
note: (skipping 4 expansions in backtrace; use -fmacro-backtrace-limit=0 to see all)
/home/brad/tmp/highway/hwy/ops/rvv-inl.h:231:3: note: expanded from macro 'HWY_RVV_FOREACH_64_EXT'
  231 |   X_MACRO(BASE, CHAR, 64, __, 32, m4, m8, m2, 2, /*MLEN=*/16, NAME, OP)
      |   ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/home/brad/tmp/highway/hwy/ops/rvv-inl.h:2463:12: note: expanded from macro 'HWY_RVV_STOREN'
 2463 |     return __riscv_v##OP##SEW##_v_##CHAR##SEW##LMUL(                           \
      |            ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
<scratch space>:126:1: note: expanded from here
  126 | __riscv_vse64_v_u64m4
      | ^~~~~~~~~~~~~~~~~~~~~
/home/brad/tmp/highway/hwy/ops/rvv-inl.h:2466:17: note: candidate function not viable: no known conversion from 'unsigned long long *' to 'unsigned long *' for 1st argument
 2466 | HWY_RVV_FOREACH(HWY_RVV_STOREN, StoreN, se, _ALL_VIRT)
      | ~~~~~~~~~~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/home/brad/tmp/highway/hwy/ops/rvv-inl.h:477:22: note: expanded from macro 'HWY_RVV_FOREACH'
  477 |   HWY_RVV_FOREACH_UI(X_MACRO, NAME, OP, LMULS)    \
      |   ~~~~~~~~~~~~~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~
/home/brad/tmp/highway/hwy/ops/rvv-inl.h:473:21: note: expanded from macro 'HWY_RVV_FOREACH_UI'
  473 |   HWY_RVV_FOREACH_U(X_MACRO, NAME, OP, LMULS)        \
      |   ~~~~~~~~~~~~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~
/home/brad/tmp/highway/hwy/ops/rvv-inl.h:461:27: note: expanded from macro 'HWY_RVV_FOREACH_U'
  461 |   HWY_RVV_FOREACH_U163264(X_MACRO, NAME, OP, LMULS)
      |   ~~~~~~~~~~~~~~~~~~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~
note: (skipping 4 expansions in backtrace; use -fmacro-backtrace-limit=0 to see all)
/home/brad/tmp/highway/hwy/ops/rvv-inl.h:231:3: note: expanded from macro 'HWY_RVV_FOREACH_64_EXT'
  231 |   X_MACRO(BASE, CHAR, 64, __, 32, m4, m8, m2, 2, /*MLEN=*/16, NAME, OP)
      |   ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/home/brad/tmp/highway/hwy/ops/rvv-inl.h:2463:12: note: expanded from macro 'HWY_RVV_STOREN'
 2463 |     return __riscv_v##OP##SEW##_v_##CHAR##SEW##LMUL(                           \
      |            ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
<scratch space>:126:1: note: expanded from here
  126 | __riscv_vse64_v_u64m4
      | ^~~~~~~~~~~~~~~~~~~~~
In file included from /home/brad/tmp/highway/hwy/per_target.cc:28:
In file included from /home/brad/tmp/highway/hwy/highway.h:714:
/home/brad/tmp/highway/hwy/ops/rvv-inl.h:2466:17: error: no matching function for call to '__riscv_vse64_v_u64m8'
 2466 | HWY_RVV_FOREACH(HWY_RVV_STOREN, StoreN, se, _ALL_VIRT)
      | ~~~~~~~~~~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/home/brad/tmp/highway/hwy/ops/rvv-inl.h:477:22: note: expanded from macro 'HWY_RVV_FOREACH'
  477 |   HWY_RVV_FOREACH_UI(X_MACRO, NAME, OP, LMULS)    \
      |   ~~~~~~~~~~~~~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~
/home/brad/tmp/highway/hwy/ops/rvv-inl.h:473:21: note: expanded from macro 'HWY_RVV_FOREACH_UI'
  473 |   HWY_RVV_FOREACH_U(X_MACRO, NAME, OP, LMULS)        \
      |   ~~~~~~~~~~~~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~
/home/brad/tmp/highway/hwy/ops/rvv-inl.h:461:27: note: expanded from macro 'HWY_RVV_FOREACH_U'
  461 |   HWY_RVV_FOREACH_U163264(X_MACRO, NAME, OP, LMULS)
      |   ~~~~~~~~~~~~~~~~~~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~
note: (skipping 3 expansions in backtrace; use -fmacro-backtrace-limit=0 to see all)
/home/brad/tmp/highway/hwy/ops/rvv-inl.h:248:3: note: expanded from macro 'HWY_RVV_FOREACH_64_ALL'
  248 |   X_MACRO(BASE, CHAR, 64, __, 32, m8, __, m4, 3, /*MLEN=*/8, NAME, OP)
      |   ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/home/brad/tmp/highway/hwy/ops/rvv-inl.h:2463:12: note: expanded from macro 'HWY_RVV_STOREN'
 2463 |     return __riscv_v##OP##SEW##_v_##CHAR##SEW##LMUL(                           \
      |            ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
<scratch space>:140:1: note: expanded from here
  140 | __riscv_vse64_v_u64m8
      | ^~~~~~~~~~~~~~~~~~~~~
/home/brad/tmp/highway/hwy/ops/rvv-inl.h:2466:17: note: candidate function not viable: no known conversion from 'unsigned long long *' to 'unsigned long *' for 1st argument
 2466 | HWY_RVV_FOREACH(HWY_RVV_STOREN, StoreN, se, _ALL_VIRT)
      | ~~~~~~~~~~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/home/brad/tmp/highway/hwy/ops/rvv-inl.h:477:22: note: expanded from macro 'HWY_RVV_FOREACH'
  477 |   HWY_RVV_FOREACH_UI(X_MACRO, NAME, OP, LMULS)    \
      |   ~~~~~~~~~~~~~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~
/home/brad/tmp/highway/hwy/ops/rvv-inl.h:473:21: note: expanded from macro 'HWY_RVV_FOREACH_UI'
  473 |   HWY_RVV_FOREACH_U(X_MACRO, NAME, OP, LMULS)        \
      |   ~~~~~~~~~~~~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~
/home/brad/tmp/highway/hwy/ops/rvv-inl.h:461:27: note: expanded from macro 'HWY_RVV_FOREACH_U'
  461 |   HWY_RVV_FOREACH_U163264(X_MACRO, NAME, OP, LMULS)
      |   ~~~~~~~~~~~~~~~~~~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~
note: (skipping 3 expansions in backtrace; use -fmacro-backtrace-limit=0 to see all)
/home/brad/tmp/highway/hwy/ops/rvv-inl.h:248:3: note: expanded from macro 'HWY_RVV_FOREACH_64_ALL'
  248 |   X_MACRO(BASE, CHAR, 64, __, 32, m8, __, m4, 3, /*MLEN=*/8, NAME, OP)
      |   ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/home/brad/tmp/highway/hwy/ops/rvv-inl.h:2463:12: note: expanded from macro 'HWY_RVV_STOREN'
 2463 |     return __riscv_v##OP##SEW##_v_##CHAR##SEW##LMUL(                           \
      |            ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
<scratch space>:140:1: note: expanded from here
  140 | __riscv_vse64_v_u64m8
      | ^~~~~~~~~~~~~~~~~~~~~
In file included from /home/brad/tmp/highway/hwy/per_target.cc:28:
In file included from /home/brad/tmp/highway/hwy/highway.h:714:
/home/brad/tmp/highway/hwy/ops/rvv-inl.h:2466:17: error: no matching function for call to '__riscv_vse64_v_u64m1'
 2466 | HWY_RVV_FOREACH(HWY_RVV_STOREN, StoreN, se, _ALL_VIRT)
      | ~~~~~~~~~~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/home/brad/tmp/highway/hwy/ops/rvv-inl.h:477:22: note: expanded from macro 'HWY_RVV_FOREACH'
  477 |   HWY_RVV_FOREACH_UI(X_MACRO, NAME, OP, LMULS)    \
      |   ~~~~~~~~~~~~~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~
/home/brad/tmp/highway/hwy/ops/rvv-inl.h:473:21: note: expanded from macro 'HWY_RVV_FOREACH_UI'
  473 |   HWY_RVV_FOREACH_U(X_MACRO, NAME, OP, LMULS)        \
      |   ~~~~~~~~~~~~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~
/home/brad/tmp/highway/hwy/ops/rvv-inl.h:461:27: note: expanded from macro 'HWY_RVV_FOREACH_U'
  461 |   HWY_RVV_FOREACH_U163264(X_MACRO, NAME, OP, LMULS)
      |   ~~~~~~~~~~~~~~~~~~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~
note: (skipping 3 expansions in backtrace; use -fmacro-backtrace-limit=0 to see all)
/home/brad/tmp/highway/hwy/ops/rvv-inl.h:275:3: note: expanded from macro 'HWY_RVV_FOREACH_64_VIRT'
  275 |   X_MACRO(BASE, CHAR, 64, __, 32, m1, m2, mf2, -1, /*MLEN=*/64, NAME, OP)
      |   ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/home/brad/tmp/highway/hwy/ops/rvv-inl.h:2463:12: note: expanded from macro 'HWY_RVV_STOREN'
 2463 |     return __riscv_v##OP##SEW##_v_##CHAR##SEW##LMUL(                           \
      |            ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
<scratch space>:154:1: note: expanded from here
  154 | __riscv_vse64_v_u64m1
      | ^~~~~~~~~~~~~~~~~~~~~
/home/brad/tmp/highway/hwy/ops/rvv-inl.h:2466:17: note: candidate function not viable: no known conversion from 'unsigned long long *' to 'unsigned long *' for 1st argument
 2466 | HWY_RVV_FOREACH(HWY_RVV_STOREN, StoreN, se, _ALL_VIRT)
      | ~~~~~~~~~~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/home/brad/tmp/highway/hwy/ops/rvv-inl.h:477:22: note: expanded from macro 'HWY_RVV_FOREACH'
  477 |   HWY_RVV_FOREACH_UI(X_MACRO, NAME, OP, LMULS)    \
      |   ~~~~~~~~~~~~~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~
/home/brad/tmp/highway/hwy/ops/rvv-inl.h:473:21: note: expanded from macro 'HWY_RVV_FOREACH_UI'
  473 |   HWY_RVV_FOREACH_U(X_MACRO, NAME, OP, LMULS)        \
      |   ~~~~~~~~~~~~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~
/home/brad/tmp/highway/hwy/ops/rvv-inl.h:461:27: note: expanded from macro 'HWY_RVV_FOREACH_U'
  461 |   HWY_RVV_FOREACH_U163264(X_MACRO, NAME, OP, LMULS)
      |   ~~~~~~~~~~~~~~~~~~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~
note: (skipping 3 expansions in backtrace; use -fmacro-backtrace-limit=0 to see all)
/home/brad/tmp/highway/hwy/ops/rvv-inl.h:275:3: note: expanded from macro 'HWY_RVV_FOREACH_64_VIRT'
  275 |   X_MACRO(BASE, CHAR, 64, __, 32, m1, m2, mf2, -1, /*MLEN=*/64, NAME, OP)
      |   ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/home/brad/tmp/highway/hwy/ops/rvv-inl.h:2463:12: note: expanded from macro 'HWY_RVV_STOREN'
 2463 |     return __riscv_v##OP##SEW##_v_##CHAR##SEW##LMUL(                           \
      |            ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
<scratch space>:154:1: note: expanded from here
  154 | __riscv_vse64_v_u64m1
      | ^~~~~~~~~~~~~~~~~~~~~
In file included from /home/brad/tmp/highway/hwy/per_target.cc:28:
In file included from /home/brad/tmp/highway/hwy/highway.h:714:
/home/brad/tmp/highway/hwy/ops/rvv-inl.h:2466:17: error: no matching function for call to '__riscv_vse64_v_i64m1'
```

--- a/clang/lib/Sema/SemaRISCV.cpp
+++ b/clang/lib/Sema/SemaRISCV.cpp
@@ -131,10 +131,24 @@ static QualType RVVType2Qual(ASTContext &Context, const RVVType *Type) {
     QT = Context.BoolTy;
     break;
   case ScalarTypeKind::SignedInteger:
-    QT = Context.getIntTypeForBitwidth(Type->getElementBitwidth(), true);
+    // getIntTypeForBitwidth() picks a type purely by matching bit width, so
+    // on LP64 targets a 64-bit element would resolve to "long" even if the
+    // target's actual int64_t is "long long" (e.g. OpenBSD). Go through the
+    // target's Int64Type so this matches int64_t/uint64_t.
+    if (Type->getElementBitwidth() == 64)
+      QT = Context.getTargetInfo().getInt64Type() == TargetInfo::SignedLong
+               ? Context.LongTy
+               : Context.LongLongTy;
+    else
+      QT = Context.getIntTypeForBitwidth(Type->getElementBitwidth(), true);
     break;
   case ScalarTypeKind::UnsignedInteger:
-    QT = Context.getIntTypeForBitwidth(Type->getElementBitwidth(), false);
+    if (Type->getElementBitwidth() == 64)
+      QT = Context.getTargetInfo().getInt64Type() == TargetInfo::SignedLong
+               ? Context.UnsignedLongTy
+               : Context.UnsignedLongLongTy;
+    else
+      QT = Context.getIntTypeForBitwidth(Type->getElementBitwidth(), false);
     break;
   case ScalarTypeKind::FloatE4M3:
   case ScalarTypeKind::FloatE5M2: {

--- a/clang/test/Sema/riscv-rvv-int64-scalar-type.c
+++ b/clang/test/Sema/riscv-rvv-int64-scalar-type.c
@@ -1,0 +1,35 @@
+// RUN: %clang_cc1 -triple riscv64-unknown-openbsd -target-feature +v -ffreestanding -fsyntax-only -verify=openbsd %s
+// RUN: %clang_cc1 -triple riscv64-none-linux-gnu -target-feature +v -ffreestanding -fsyntax-only -verify=linux %s
+
+// REQUIRES: riscv-registered-target
+
+// RVV intrinsics with a 64-bit scalar/pointer operand (e.g. vse64) must use
+// the target's actual uint64_t/int64_t type, not always "unsigned long":
+// OpenBSD defines uint64_t as "unsigned long long" on every architecture,
+// while riscv64-linux (LP64) defines it as "unsigned long".
+
+#include <stdint.h>
+#include <riscv_vector.h>
+
+// uint64_t* must be accepted on every target, regardless of whether the
+// target's uint64_t happens to be "unsigned long" or "unsigned long long".
+void test_uint64_ok(uint64_t *p, vuint64m1_t v, size_t vl) {
+  __riscv_vse64_v_u64m1(p, v, vl);
+}
+
+// "unsigned long *" is only the right type for the pointee on riscv64-linux
+// (where uint64_t is "unsigned long"); on OpenBSD, uint64_t is
+// "unsigned long long", so this should be an incompatible pointer type.
+void test_unsigned_long(unsigned long *p, vuint64m1_t v, size_t vl) {
+  __riscv_vse64_v_u64m1(p, v, vl);
+  // openbsd-error@-1 {{incompatible pointer types passing 'unsigned long *' to parameter of type 'unsigned long long *'}}
+  // openbsd-note@-2 {{passing argument to parameter here}}
+}
+
+// Conversely, "unsigned long long *" is only correct on OpenBSD; on
+// riscv64-linux, uint64_t is "unsigned long", so this should fail there.
+void test_unsigned_long_long(unsigned long long *p, vuint64m1_t v, size_t vl) {
+  __riscv_vse64_v_u64m1(p, v, vl);
+  // linux-error@-1 {{incompatible pointer types passing 'unsigned long long *' to parameter of type 'unsigned long *'}}
+  // linux-note@-2 {{passing argument to parameter here}}
+}

